File: Solutions/Python/Problem 004.py
# Problem 4 - Largest palindrome product
# https://projecteuler.net/problem=4
# Answer = 906609

def palindrome():
    """
    Find the largest palindrome product of two 3-digit numbers using nested loops.

    Time Complexity: O(n^2), where n is the number of 3-digit numbers (900).

    Returns:
        int: The largest palindrome product.
    """
    palindromes = []
    for i in range(100, 1000):
        for j in range(100, 1000):
            product = i * j
            if str(product) == str(product)[::-1]:
                palindromes.append(product)
    return max(palindromes)

def main():
    largest_palindrome = palindrome()
    print(f"The largest palindrome product of two 3-digit numbers is {largest_palindrome}.")

if __name__ == "__main__":
    main()

File: Solutions/Rust/problem_004.rs
// Problem 4 - Largest palindrome product
// https://projecteuler.net/problem=4
// Answer = 906609

fn question() {
    println!("Find the largest palindrome made from the product of two 3-digit numbers.");
}

fn is_palindrome(num:i32) -> bool{
    let mut reverse_num = 0;
    let mut temp_num = num;
    while temp_num != 0{
        reverse_num = (reverse_num * 10) + (temp_num % 10);
        temp_num /= 10;
    }
    reverse_num == num
}

fn find_largest_palindrome() -> i32 {
    // Finds the largest palindrome made from the product of two 3-digit numbers.
    // Time complexity: O(n^2), where 'n' is the range (100 to 999) for the numbers being multiplied.
    let mut largest_palindrome = 0;

    for i in (100..1000).rev() {
        if i * 1000 < largest_palindrome{
            break
        }
        for j in (i..1000).rev() {
            let product = i * j;
            if product < largest_palindrome{
                break;
            }
            if is_palindrome(product) {
                largest_palindrome = product;
            }
        }
    }
    largest_palindrome
}

fn main() {
    question();
    let result = find_largest_palindrome();
    println!("The largest palindrome is: {}", result);
}

